builder/simplify.cc

```diff
@@ -783,23 +764,80 @@ class simplifier : public node_mutator {
   bool prove_false(const expr& e) { return prove_constant_false(where_true(e).max); }
 
   void visit(const variable* op) override {
-    if (vars.contains(op->sym)) {
-      expr_info info = *vars[op->sym];
-      if (info.replacement.defined()) {
-        // TODO: This seems like it might be expensive, but it's the simplest way to get correct bounds and alignment
-        // information.
-        // TODO: Maybe we should intersect any information we already had with this?
-        mutate_and_set_result(info.replacement);
-      } else if (auto c = as_constant(info.bounds.as_point())) {
-        set_result(info.bounds.min, {point(info.bounds.min), alignment_type()});
-      } else {
-        if (!info.bounds.min.defined()) info.bounds.min = expr(op);
-        if (!info.bounds.max.defined()) info.bounds.max = expr(op);
-        set_result(op, std::move(info));
+    if (op->field != buffer_field::none) {
+      var new_sym = op->sym;
+      if (vars.contains(op->sym)) {
+        const expr_info& info = *vars[op->sym];
+        if (info.replacement.defined()) {
+          new_sym = info.replacement_sym();
+        }
+      }
+      const std::optional<buffer_info>& info = buffers[new_sym];
+      expr result = new_sym == op->sym ? expr(op) : variable::make(new_sym, op->field, op->dim);
+      expr bounds = result;
+      if (info) {
+        // TODO: We substitute here because we can't prove things like buffer_elem_size(x) == buffer_elem_size(y) where
+        // x is a crop of y. If we can fix that, we don't need to substitute here, which seems better.
+        auto visit_buffer_meta_value = [&, this](expr x) {
+          // There are many conditions in which we should substitute buffer meta:
+          // - We're being asked to substitute it (decl is undefined).
+          // - The value is something we should substitute (it's simple and pure).
+          // - The value is another buffer meta expression.
+          // - We're trying to prove something (as opposed to producing a simplified expression).
+          if (!info->decl.defined() || should_substitute(x) || x.as<variable>() || (proving && x.defined())) {
+            if (!match(x, op)) {
+              // This is a value we should substitute, and it's different from what we started with.
+              mutate_and_set_result(x);
+              return true;
+            }
+          }
+          if (x.defined()) {
+            bounds = x;
+          }
+          return false;
+        };
+        switch (op->field) {
+        case buffer_field::elem_size:
+          if (visit_buffer_meta_value(info->elem_size)) return;
+          break;
+        case buffer_field::min:
+        case buffer_field::max:
+        case buffer_field::stride:
+        case buffer_field::fold_factor:
+          if (op->dim < static_cast<index_t>(info->dims.size())) {
```

Comment:
Should this (also) be an assert()?

Reply:
No, we might not know about all the dimensions of a buffer.